CTE tests throw error with recursive tests

```shell
sqlmesh --version
0.131.0
```

When the test of the following model includes cte tests, an error is thrown. The model itself works perfectly well on duckdb, to which the error is related, suggesting that the error is not directly related to duckdb.

Model:
```sql
MODEL (
  name sandbox.recursive_cte_bug,
  kind FULL
);

WITH
    RECURSIVE
    chained_contacts AS (
        -- Start with the initial set of contacts and their immediate nodes
        SELECT
            id_contact_a,
            id_contact_b
        FROM
            sandbox.x420_source

        UNION ALL

        -- Recursive step to find further connected nodes
        SELECT
            chained_contacts.id_contact_a,
            unfactorized_duplicates.id_contact_b
        FROM
            chained_contacts
                JOIN sandbox.x420_source AS unfactorized_duplicates
                     ON chained_contacts.id_contact_b = unfactorized_duplicates.id_contact_a
    ),
    id_contact_a_with_aggregated_id_contact_bs AS (
        SELECT
            id_contact_a,
            ARRAY_AGG(DISTINCT id_contact_b ORDER BY id_contact_b) AS aggregated_id_contact_bs
        FROM
            chained_contacts
        GROUP BY
            id_contact_a
    )
SELECT
    ARRAY_CONCAT([id_contact_a], aggregated_id_contact_bs) AS aggregated_duplicates
FROM
    id_contact_a_with_aggregated_id_contact_bs
WHERE
    id_contact_a NOT IN (
        SELECT DISTINCT
            id_contact_b
        FROM
            sandbox.x420_source
    )
ORDER BY
    id_contact_a
```

When running the following test:
```yaml
reproduce_bug:
  model: sandbox.recursive_cte_bug
  inputs:
    sandbox.x420_source:
      rows:
        - id_contact_a: "a"
          id_contact_b: "b"
        - id_contact_a: "b"
          id_contact_b: "c"
        - id_contact_a: "c"
          id_contact_b: "d"
        - id_contact_a: "a"
          id_contact_b: "g"
        - id_contact_a: "b"
          id_contact_b: "e"
        - id_contact_a: "c"
          id_contact_b: "f"
        - id_contact_a: "x"
          id_contact_b: "y"
  outputs:
    ctes:
      id_contact_a_with_aggregated_id_contact_bs:
        - id_contact_a: a
          aggregated_id_contact_bs:
            - b
            - c
            - d
            - e
            - f
            - g
        - id_contact_a: x
          aggregated_id_contact_bs:
            - y
    query:
      rows:
        - aggregated_duplicates:
          - a
          - b
          - c
          - d
          - e
          - f
          - g
        - aggregated_duplicates:
          - x
          - y
```

An error is thrown when running test:
```shell
sqlmesh test
.......E..
======================================================================
ERROR: reproduce_bug (/project-dir/tests/test_recursive_cte_bug.yaml) (cte='"id_contact_a_with_aggregated_id_contact_bs"')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/project-dir/.venv/lib/python3.12/site-packages/sqlmesh/core/test/definition.py", line 544, in test_ctes
    actual = self._execute(cte_query)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/project-dir/.venv/lib/python3.12/site-packages/sqlmesh/core/test/definition.py", line 491, in _execute
    return self.engine_adapter.fetchdf(query)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/project-dir/.venv/lib/python3.12/site-packages/sqlmesh/core/engine_adapter/base.py", line 1932, in fetchdf
    df = self._fetch_native_df(query, quote_identifiers=quote_identifiers)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/project-dir/.venv/lib/python3.12/site-packages/sqlmesh/core/engine_adapter/base.py", line 1925, in _fetch_native_df
    self.execute(query, quote_identifiers=quote_identifiers)
  File "/project-dir/.venv/lib/python3.12/site-packages/sqlmesh/core/engine_adapter/base.py", line 2059, in execute
    self._execute(sql, **kwargs)
  File "/project-dir/.venv/lib/python3.12/site-packages/sqlmesh/core/engine_adapter/base.py", line 2065, in _execute
    self.cursor.execute(sql, **kwargs)
duckdb.duckdb.BinderException: Binder Error: Circular reference to CTE "chained_contacts", There are two possible solutions. 
1. use WITH RECURSIVE to use recursive CTEs. 
2. If you want to use the TABLE name "chained_contacts" the same as the CTE name, please explicitly add "SCHEMA" before table name. You can try "main.chained_contacts" (main is the duckdb default schema)

----------------------------------------------------------------------
Ran 10 tests in 0.168s

FAILED (errors=1)
```

The error does not occur when removing the CTE test:
```yaml
reproduce_bug:
  model: sandbox.recursive_cte_bug
  inputs:
    sandbox.x420_source:
      rows:
        - id_contact_a: "a"
          id_contact_b: "b"
        - id_contact_a: "b"
          id_contact_b: "c"
        - id_contact_a: "c"
          id_contact_b: "d"
        - id_contact_a: "a"
          id_contact_b: "g"
        - id_contact_a: "b"
          id_contact_b: "e"
        - id_contact_a: "c"
          id_contact_b: "f"
        - id_contact_a: "x"
          id_contact_b: "y"
  outputs:
    query:
      rows:
        - aggregated_duplicates:
          - a
          - b
          - c
          - d
          - e
          - f
          - g
        - aggregated_duplicates:
          - x
          - y
```

test run output:
```shell
sqlmesh test
..........
----------------------------------------------------------------------
Ran 10 tests in 0.168s

OK
```

Actually, why does the tracebacks sometimes show and sometimes not? I think it would be better is they did not.

As a repeat of the results in #3352:
```shell
sqlmesh test
...F......
======================================================================
FAIL: test_arrays_bug (/project-dir/tests/test_arrays_bug.yaml)
----------------------------------------------------------------------
AssertionError: Data mismatch (exp: expected, act: actual)

  aggregated_duplicates        
                    exp     act
0                (c, b)  (b, c)

----------------------------------------------------------------------
Ran 10 tests in 0.159s

FAILED (failures=1)
```

but for this one (with my initially wrong test):
```shell
sqlmesh test
........F..
======================================================================
FAIL: reproduce_bug (/project-dir/sqlmesh-20-duplicate-arbitrage/tests/test_recursive_cte_bug.yaml) (cte='"id_contact_a_with_aggregated_id_contact_bs"')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "project-dir/sqlmesh/core/test/definition.py", line 550, in test_ctes
    self.assert_equal(expected, actual, sort=sort, partial=partial)
  File "project-dir/sqlmesh/core/test/definition.py", line 273, in assert_equal
    raise e
  File "project-dir/sqlmesh/core/test/definition.py", line 247, in assert_equal
    pd.testing.assert_frame_equal(
  File "project-dir/.venv/lib/python3.12/site-packages/pandas/_testing/asserters.py", line 1218, in assert_frame_equal
    raise_assert_detail(
  File "project-dir/.venv/lib/python3.12/site-packages/pandas/_testing/asserters.py", line 614, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: Data mismatch (rows are different)

Unexpected rows:

  id_contact_a aggregated_id_contact_bs
0            b             (c, d, e, f)
1            c                   (d, f)

----------------------------------------------------------------------
Ran 11 tests in 0.176s

FAILED (failures=1)
```

--- a/docs/reference/cli.md
+++ b/docs/reference/cli.md
@@ -92,7 +92,7 @@ Usage: sqlmesh create_test [OPTIONS] MODEL
 
 Options:
   -q, --query <TEXT TEXT>...  Queries that will be used to generate data for
-                              the model's dependencies.  [required]
+                              the model's dependencies.
   -o, --overwrite             When true, the fixture file will be overwritten
                               in case it already exists.
   -v, --var <TEXT TEXT>...    Key-value pairs that will define variables

--- a/docs/reference/notebook.md
+++ b/docs/reference/notebook.md
@@ -366,8 +366,9 @@ options:
 
 #### create_test
 ```
-%create_test --query QUERY [QUERY ...] [--overwrite]
+%create_test [--query QUERY [QUERY ...]] [--overwrite]
                    [--var VAR [VAR ...]] [--path PATH] [--name NAME]
+                   [--include-ctes]
                    model
 
 Generate a unit test fixture for a given model.

--- a/sqlmesh/cli/main.py
+++ b/sqlmesh/cli/main.py
@@ -530,7 +530,7 @@ def dag(ctx: click.Context, file: str, select_model: t.List[str]) -> None:
     "queries",
     type=(str, str),
     multiple=True,
-    required=True,
+    default=[],
     help="Queries that will be used to generate data for the model's dependencies.",
 )
 @click.option(

--- a/sqlmesh/core/test/definition.py
+++ b/sqlmesh/core/test/definition.py
@@ -5,6 +5,7 @@ import typing as t
 import unittest
 from collections import Counter
 from contextlib import AbstractContextManager, nullcontext
+from itertools import chain
 from pathlib import Path
 from unittest.mock import patch
 
@@ -126,25 +127,25 @@ class ModelTest(unittest.TestCase):
 
         for name, values in self.body.get("inputs", {}).items():
             all_types_are_known = False
-            known_columns_to_types: t.Dict[str, exp.DataType] = {}
+            columns_to_known_types: t.Dict[str, exp.DataType] = {}
 
             model = self.models.get(name)
             if model:
                 inferred_columns_to_types = model.columns_to_types or {}
-                known_columns_to_types = {
+                columns_to_known_types = {
                     c: t for c, t in inferred_columns_to_types.items() if type_is_known(t)
                 }
                 all_types_are_known = bool(inferred_columns_to_types) and (
-                    len(known_columns_to_types) == len(inferred_columns_to_types)
+                    len(columns_to_known_types) == len(inferred_columns_to_types)
                 )
 
             # Types specified in the test will override the corresponding inferred ones
-            known_columns_to_types.update(values.get("columns", {}))
+            columns_to_known_types.update(values.get("columns", {}))
 
             rows = values.get("rows")
             if not all_types_are_known and rows:
                 for col, value in rows[0].items():
-                    if col not in known_columns_to_types:
+                    if col not in columns_to_known_types:
                         v_type = annotate_types(exp.convert(value)).type or type(value).__name__
                         v_type = exp.maybe_parse(
                             v_type, into=exp.DataType, dialect=self._test_adapter_dialect
@@ -159,21 +160,21 @@ class ModelTest(unittest.TestCase):
                                 self.path,
                             )
 
-                        known_columns_to_types[col] = v_type
+                        columns_to_known_types[col] = v_type
 
             if rows is None:
                 query_or_df: exp.Query | pd.DataFrame = self._add_missing_columns(
-                    values["query"], known_columns_to_types
+                    values["query"], columns_to_known_types
                 )
-                if known_columns_to_types:
-                    known_columns_to_types = {
-                        col: known_columns_to_types[col] for col in query_or_df.named_selects
+                if columns_to_known_types:
+                    columns_to_known_types = {
+                        col: columns_to_known_types[col] for col in query_or_df.named_selects
                     }
             else:
-                query_or_df = self._create_df(values, columns=known_columns_to_types)
+                query_or_df = self._create_df(values, columns=columns_to_known_types)
 
             self.engine_adapter.create_view(
-                self._test_fixture_table(name), query_or_df, known_columns_to_types
+                self._test_fixture_table(name), query_or_df, columns_to_known_types
             )
 
     def tearDown(self) -> None:
@@ -525,7 +526,7 @@ class ModelTest(unittest.TestCase):
 
 
 class SqlModelTest(ModelTest):
-    def test_ctes(self, ctes: t.Dict[str, exp.Expression]) -> None:
+    def test_ctes(self, ctes: t.Dict[str, exp.Expression], recursive: bool = False) -> None:
         """Run CTE queries and compare output to expected output"""
         for cte_name, values in self.body["outputs"].get("ctes", {}).items():
             with self.subTest(cte=cte_name):
@@ -535,11 +536,13 @@ class SqlModelTest(ModelTest):
                     )
 
                 cte_query = ctes[cte_name].this
-                for alias, cte in ctes.items():
-                    cte_query = cte_query.with_(alias, cte.this)
 
-                partial = values.get("partial")
                 sort = cte_query.args.get("order") is None
+                partial = values.get("partial")
+
+                cte_query = exp.select(*_projection_identifiers(cte_query)).from_(cte_name)
+                for alias, cte in ctes.items():
+                    cte_query = cte_query.with_(alias, cte.this, recursive=recursive)
 
                 actual = self._execute(cte_query)
                 expected = self._create_df(values, columns=cte_query.named_selects, partial=partial)
@@ -548,13 +551,16 @@ class SqlModelTest(ModelTest):
 
     def runTest(self) -> None:
         query = self._render_model_query()
+        with_clause = query.args.get("with")
 
-        self.test_ctes(
-            {
-                self._normalize_model_name(cte.alias, with_default_catalog=False): cte
-                for cte in query.ctes
-            }
-        )
+        if with_clause:
+            self.test_ctes(
+                {
+                    self._normalize_model_name(cte.alias, with_default_catalog=False): cte
+                    for cte in query.ctes
+                },
+                recursive=with_clause.recursive,
+            )
 
         values = self.body["outputs"].get("query")
         if values is not None:
@@ -732,14 +738,23 @@ def generate_test(
     if isinstance(model, SqlModel):
         assert isinstance(test, SqlModelTest)
         model_query = test._render_model_query()
+        with_clause = model_query.args.get("with")
 
-        if include_ctes:
+        if with_clause and include_ctes:
             ctes = {}
+            recursive = with_clause.recursive
             previous_ctes: t.List[exp.CTE] = []
+
             for cte in model_query.ctes:
                 cte_query = cte.this
-                for prev in previous_ctes:
-                    cte_query = cte_query.with_(prev.alias, prev.this)
+                cte_identifier = cte.args["alias"].this
+
+                cte_query = exp.select(*_projection_identifiers(cte_query)).from_(cte_identifier)
+
+                for prev in chain(previous_ctes, [cte]):
+                    cte_query = cte_query.with_(
+                        prev.args["alias"].this, prev.this, recursive=recursive
+                    )
 
                 cte_output = test._execute(cte_query)
                 ctes[cte.alias] = (
@@ -773,6 +788,19 @@ def generate_test(
     fixture_path.parent.mkdir(exist_ok=True, parents=True)
     with open(fixture_path, "w", encoding="utf-8") as file:
         yaml.dump({test_name: test_body}, file)
+
+
+def _projection_identifiers(query: exp.Query) -> t.List[str | exp.Identifier]:
+    identifiers: t.List[str | exp.Identifier] = []
+    for select in query.selects:
+        if isinstance(select, exp.Alias):
+            identifiers.append(select.args["alias"])
+        elif isinstance(select, exp.Column):
+            identifiers.append(select.this)
+        else:
+            identifiers.append(select.output_name)
+
+    return identifiers
 
 
 def _raise_if_unexpected_columns(

--- a/sqlmesh/magics.py
+++ b/sqlmesh/magics.py
@@ -858,7 +858,7 @@ class SQLMeshMagics(Magics):
         "-q",
         type=str,
         nargs="+",
-        required=True,
+        default=[],
         help="Queries that will be used to generate data for the model's dependencies.",
     )
     @argument(

--- a/tests/core/test_test.py
+++ b/tests/core/test_test.py
@@ -1450,12 +1450,12 @@ test_foo:
         )
 
 
-def test_pyspark_python_model() -> None:
+def test_pyspark_python_model(tmp_path: Path) -> None:
     spark_connection_config = SparkConnectionConfig(
         config={
             "spark.master": "local",
-            "spark.sql.warehouse.dir": "/tmp/data_dir",
-            "spark.driver.extraJavaOptions": "-Dderby.system.home=/tmp/derby_dir",
+            "spark.sql.warehouse.dir": f"{tmp_path}/data_dir",
+            "spark.driver.extraJavaOptions": f"-Dderby.system.home={tmp_path}/derby_dir",
         },
     )
     config = Config(
@@ -1569,6 +1569,100 @@ test_foo:
             call('SELECT 1 AS "a"'),
             call('DROP SCHEMA IF EXISTS "memory"."my_schema" CASCADE'),
         ]
+    )
+
+
+def test_complicated_recursive_cte() -> None:
+    model_sql = """
+WITH
+    RECURSIVE
+    chained_contacts AS (
+        -- Start with the initial set of contacts and their immediate nodes
+        SELECT
+            id_contact_a,
+            id_contact_b
+        FROM
+            source
+
+        UNION ALL
+
+        -- Recursive step to find further connected nodes
+        SELECT
+            chained_contacts.id_contact_a,
+            unfactorized_duplicates.id_contact_b
+        FROM
+            chained_contacts
+                JOIN source AS unfactorized_duplicates
+                     ON chained_contacts.id_contact_b = unfactorized_duplicates.id_contact_a
+    ),
+    id_contact_a_with_aggregated_id_contact_bs AS (
+        SELECT
+            id_contact_a,
+            ARRAY_AGG(DISTINCT id_contact_b ORDER BY id_contact_b) AS aggregated_id_contact_bs
+        FROM
+            chained_contacts
+        GROUP BY
+            id_contact_a
+    )
+SELECT
+    ARRAY_CONCAT([id_contact_a], aggregated_id_contact_bs) AS aggregated_duplicates
+FROM
+    id_contact_a_with_aggregated_id_contact_bs
+WHERE
+    id_contact_a NOT IN (
+        SELECT DISTINCT
+            id_contact_b
+        FROM
+            source
+    )
+ORDER BY
+    id_contact_a
+    """
+
+    _check_successful_or_raise(
+        _create_test(
+            body=load_yaml(
+                """
+test_recursive_ctes:
+  model: test
+  inputs:
+    source:
+      rows:
+        - id_contact_a: "a"
+          id_contact_b: "b"
+        - id_contact_a: "b"
+          id_contact_b: "c"
+        - id_contact_a: "c"
+          id_contact_b: "d"
+        - id_contact_a: "a"
+          id_contact_b: "g"
+        - id_contact_a: "b"
+          id_contact_b: "e"
+        - id_contact_a: "c"
+          id_contact_b: "f"
+        - id_contact_a: "x"
+          id_contact_b: "y"
+  outputs:
+    ctes:
+      id_contact_a_with_aggregated_id_contact_bs:
+        - id_contact_a: a
+          aggregated_id_contact_bs: [b, c, d, e, f, g]
+        - id_contact_a: x
+          aggregated_id_contact_bs: [y]
+        - id_contact_a: b
+          aggregated_id_contact_bs: [c, d, e, f]
+        - id_contact_a: c
+          aggregated_id_contact_bs: [d, f]
+    query:
+      rows:
+        - aggregated_duplicates: [a, b, c, d, e, f, g]
+        - aggregated_duplicates: [x, y]
+                """
+            ),
+            test_name="test_recursive_ctes",
+            model=_create_model(model_sql),
+            context=Context(config=Config(model_defaults=ModelDefaultsConfig(dialect="duckdb"))),
+        ).run()
     )
 
 
@@ -1789,3 +1883,35 @@ def test_test_generation_with_decimal(tmp_path: Path, mocker: MockerFixture) -> 
     assert "test_foo" in test
     assert test["test_foo"]["inputs"] == {"sqlmesh_example.bar": [{"dec_col": "1.23"}]}
     assert test["test_foo"]["outputs"] == {"query": [{"dec_col": "1.23"}]}
+
+
+def test_test_generation_with_recursive_ctes(tmp_path: Path) -> None:
+    init_example_project(tmp_path, dialect="duckdb")
+
+    config = Config(
+        default_connection=DuckDBConnectionConfig(),
+        model_defaults=ModelDefaultsConfig(dialect="duckdb"),
+    )
+    foo_sql_file = tmp_path / "models" / "foo.sql"
+    foo_sql_file.write_text(
+        "MODEL (name sqlmesh_example.foo);"
+        "WITH RECURSIVE t AS (SELECT 1 AS c UNION ALL SELECT c + 1 FROM t WHERE c < 3) SELECT c FROM t"
+    )
+
+    context = Context(paths=tmp_path, config=config)
+    context.plan(auto_apply=True)
+
+    context.create_test("sqlmesh_example.foo", input_queries={}, overwrite=True, include_ctes=True)
+
+    test = load_yaml(context.path / c.TESTS / "test_foo.yaml")
+    assert len(test) == 1
+    assert "test_foo" in test
+    assert test["test_foo"]["inputs"] == {}
+    assert test["test_foo"]["outputs"] == {
+        "query": [{"c": 1}, {"c": 2}, {"c": 3}],
+        "ctes": {
+            "t": [{"c": 1}, {"c": 2}, {"c": 3}],
+        },
+    }
+
+    _check_successful_or_raise(context.test())

